src/packages/cli/src/detach.ts

```diff
@@ -0,0 +1,261 @@
+      process.exitCode = 1;
+      reject(err);
+    });
+
+    child.on("exit", (code: number) => {
+      // This shouldn't happen, so ensure that we surface a non-zero exit code.
+      process.exitCode = code === 0 ? 1 : code;
+      reject(
+        new Error(
+          `${START_ERROR}\nThe detached instance exited with error code: ${code}`
+        )
+      );
+    });
+  });
+
+  // destroy the ReadableStream exposed by the child process, to allow the
+  // parent to exit gracefully.
+  child.stderr.destroy();
+  child.unref();
+  child.disconnect();
+
+  const flavor = instanceInfo.flavor;
+  const { host, port } = instanceInfo.server;
+  const cmd =
+    process.platform === "win32"
+      ? path.basename(process.execPath)
+      : [process.execPath, ...process.execArgv, module, ...childArgs].join(" ");
+
+  const pid = child.pid;
+  const startTime = Date.now();
+
+  const instance: DetachedInstance = {
+    startTime,
+    pid,
+    instanceName: createInstanceName(),
+    host,
+    port,
+    flavor,
+    cmd,
+    version
+  };
+
+  while (true) {
+    const instanceFilename = await getInstanceFilePath(instance.instanceName);
+    try {
+      await writeFile(instanceFilename, JSON.stringify(instance), {
+        // wx means "Open file for writing, but fail if the path exists". see
+        // https://nodejs.org/api/fs.html#file-system-flags
+        flag: "wx"
+      });
+      break;
+    } catch (err) {
+      switch ((err as CodedError).code) {
+        case "EEXIST":
+          // an instance already exists with this name
+          instance.instanceName = createInstanceName();
+          break;
+        case "ENOENT":
+          // we don't check whether the folder exists before writing, as that's
+          // a very uncommon case. Catching the exception and subsequently
+          // creating the directory is faster in the majority of cases.
+          await mkdir(dataPath, { recursive: true });
+          break;
+        default:
+          throw err;
+      }
+    }
+  }
+
+  return instance;
+}
+
+/**
+ * Fetch all instance of Ganache running in detached mode. Cleans up any
+ * instance files for processes that are no longer running.
+ * @returns {Promise<DetachedInstance[]>} resolves with an array of instances
+ */
+export async function getDetachedInstances(): Promise<DetachedInstance[]> {
+  let files: string[];
+  let processes: ProcessDescriptor[];
+  try {
+    [files, processes] = await Promise.all([readdir(await dataPath), psList()]);
+  } catch (err) {
+    if ((err as CodedError).code !== "ENOENT") {
+      throw err;
+    }
+    return [];
+  }
+  const instances: DetachedInstance[] = [];
+
+  for (let i = 0; i < files.length; i++) {
+    const instanceName = files[i];
+    let shouldRemoveFile = false;
+
+    try {
+      // getDetachedInstanceByName() throws if the instance file is not found or
+      // cannot be parsed
+      const instance = await getDetachedInstanceByName(instanceName);
```

Comment:
Is there a reason not to parallelize this loop?

Reply:
Nope - I've updated this.